weaver/processes/wps3_process.py

```diff
@@ -227,113 +216,98 @@ def prepare(self):
+        # type: (str) -> bool
+        job_status_uri = monitor_reference
+        job_status_data = self.get_job_status(job_status_uri)
+        job_status_value = status.map_status(job_status_data["status"])
+        job_id = job_status_data["jobID"]
 
         self.update_status("Monitoring job on remote ADES : {0}".format(job_status_uri),
                            REMOTE_JOB_PROGRESS_MONITORING, status.STATUS_RUNNING)
 
         while job_status_value not in status.JOB_STATUS_CATEGORIES[status.JOB_STATUS_CATEGORY_FINISHED]:
             sleep(5)
-            job_status = self.get_job_status(job_status_uri)
-            job_status_value = status.map_status(job_status["status"])
+            job_status_data = self.get_job_status(job_status_uri)
+            job_status_value = status.map_status(job_status_data["status"])
 
             LOGGER.debug(get_log_monitor_msg(job_id, job_status_value,
-                                             job_status.get("percentCompleted", 0),
-                                             get_any_message(job_status), job_status.get("statusLocation")))
+                                             job_status_data.get("percentCompleted", 0),
+                                             get_any_message(job_status_data), job_status_data.get("statusLocation")))
             self.update_status(get_job_log_msg(status=job_status_value,
-                                               message=get_any_message(job_status),
-                                               progress=job_status.get("percentCompleted", 0),
-                                               duration=job_status.get("duration", None)),  # get if available
-                               map_progress(job_status.get("percentCompleted", 0),
+                                               message=get_any_message(job_status_data),
+                                               progress=job_status_data.get("percentCompleted", 0),
+                                               duration=job_status_data.get("duration", None)),  # get if available
+                               map_progress(job_status_data.get("percentCompleted", 0),
                                             REMOTE_JOB_PROGRESS_MONITORING, REMOTE_JOB_PROGRESS_FETCH_OUT),
                                status.STATUS_RUNNING)
 
         if job_status_value != status.STATUS_SUCCEEDED:
             LOGGER.debug(get_log_monitor_msg(job_id, job_status_value,
-                                             job_status.get("percentCompleted", 0),
-                                             get_any_message(job_status), job_status.get("statusLocation")))
-            raise Exception(job_status)
+                                             job_status_data.get("percentCompleted", 0),
+                                             get_any_message(job_status_data), job_status_data.get("statusLocation")))
+            raise PackageExecutionError(job_status_data)
         return job_id
```

Comment:
Should return a bool  as defined in `WpsProcessInterface.monitor`. 
Is there a reason why the monitoring loop is not implemented in the base with implementations only providing a function to get the current status. I see a lot of differences between wps1 and wps3 implementations on top of the bad return values. For example wps3 has a sleep of 5 seconds, while wps1 seems to query the server as fast as possible. Also, there is exception handling in wps1 with some retries while wps3 will raise as soon as something goes wrong.

Reply:
Good catch. Should indeed by a bool.

The reason they are different is simply because I didn't try to completely rewrite the loops too much, but only move the existing code in a more standard format focusing on the stage-in/out mostly. 

I agree that another (better) refactoring step would be to have an abstract `get_status` and standardize the monitoring itself. I might give it a try in a following PR.

Reply:
for reference https://github.com/crim-ca/weaver/issues/395